core_types/src/eav.rs

```diff
@@ -122,88 +176,117 @@ pub trait EntityAttributeValueStorage: objekt::Clone + Send + Sync + Debug {
+        Ok(Some(new_eav.clone()))
     }
 
-    fn unthreadable_fetch_eav(
+    fn fetch_eav(
         &self,
         entity: Option<Entity>,
         attribute: Option<Attribute>,
         value: Option<Value>,
-    ) -> Result<HashSet<EntityAttributeValue>, HolochainError> {
-        let filtered = self
-            .storage
-            .iter()
-            .cloned()
-            .filter(|eav| match entity {
-                Some(ref e) => &eav.entity() == e,
-                None => true,
+        index_query: IndexQuery,
+    ) -> Result<BTreeSet<EntityAttributeValueIndex>, HolochainError> {
+        let map = self.storage.read()?;
+        let new_map = map.clone();
+        let filtered = new_map
+            .into_iter()
+            .filter(|e| EntityAttributeValueIndex::filter_on_eav(&e.entity(), entity.as_ref()))
+            .filter(|e| {
+                EntityAttributeValueIndex::filter_on_eav(&e.attribute(), attribute.as_ref())
             })
-            .filter(|eav| match attribute {
-                Some(ref a) => &eav.attribute() == a,
-                None => true,
+            .filter(|e| EntityAttributeValueIndex::filter_on_eav(&e.value(), value.as_ref()))
+            .filter(|e| {
+                index_query
+                    .start_time()
+                    .map(|start| start <= e.index())
+                    .unwrap_or_else(|| {
+                        let latest = get_latest(e.clone(), map.clone())
+                            .unwrap_or(EntityAttributeValueIndex::default());
+                        latest.index() == e.index()
+                    })
             })
-            .filter(|eav| match value {
-                Some(ref v) => &eav.value() == v,
-                None => true,
+            .filter(|e| {
+                index_query
+                    .end_time()
+                    .map(|end| end >= e.index())
+                    .unwrap_or_else(|| {
+                        let latest = get_latest(e.clone(), map.clone())
+                            .unwrap_or(EntityAttributeValueIndex::default());
+                        latest.index() == e.index()
+                    })
             })
-            .collect::<HashSet<EntityAttributeValue>>();
-        Ok(filtered)
-    }
-}
+            .collect::<BTreeSet<EntityAttributeValueIndex>>();
 
-impl PartialEq for EntityAttributeValueStorage {
-    fn eq(&self, other: &EntityAttributeValueStorage) -> bool {
-        self.fetch_eav(None, None, None) == other.fetch_eav(None, None, None)
+        Ok(filtered)
     }
 }
 
-#[derive(Clone, Debug)]
-pub struct ExampleEntityAttributeValueStorage {
-    content: Arc<RwLock<ExampleEntityAttributeValueStorageNonSync>>,
-}
-
-impl ExampleEntityAttributeValueStorage {
-    pub fn new() -> HcResult<ExampleEntityAttributeValueStorage> {
-        Ok(ExampleEntityAttributeValueStorage {
-            content: Arc::new(RwLock::new(ExampleEntityAttributeValueStorageNonSync::new())),
+pub fn get_latest(
```

Comment:
can get_latest just be fetch_eav with take 1 or first or last?

is this duplicated logic or different for some reason?

Reply:
This is obtaining get latest for each and every entry in the list, it looks duplicated because it is being called while doing an fetch_eav. what we need is a fetch_eav_lazy which gets you the iterator so that you can be able to perform operations on it but that is more of a follow up. We are getting the latest timestamp of each entry that passes through that iterator